File: 705/design-hashset.js
/**
 * Initialize your data structure here.
 */
var MyHashSet = function () {
    this.buckets = 1000;
    this.data = [];
    for (let i = 0; i < this.buckets; i++) {
        this.data.push([]);
    }
    // this.data = new Array(this.buckets).map(e => []);
};

MyHashSet.prototype.hash = function (key) {
    return key % this.buckets;
};

MyHashSet.prototype.pos = function (key) {
    return (key - this.hash(key)) / this.buckets;
};

/** 
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.add = function (key) {
    let hashKey = this.hash(key);
    if (this.data[hashKey].length == 0) {
        for (let i = 0; i <= this.buckets; i++) {
            this.data[hashKey].push(0);
        }
    }
    this.data[hashKey][this.pos(key)] = 1;
    return;
};

/** 
 * @param {number} key
 * @return {void}
 */
MyHashSet.prototype.remove = function (key) {
    let hashKey = this.hash(key);
    if (this.data[hashKey].length)
        this.data[hashKey][this.pos(key)] = 0;
    return;
};

/**
 * Returns true if this set did not already contain the specified element 
 * @param {number} key
 * @return {boolean}
 */
MyHashSet.prototype.contains = function (key) {
    let hashKey = this.hash(key);
    return this.data[hashKey].length != 0 && this.data[hashKey][this.pos(key)] == 1;
};

/** 
 * Your MyHashSet object will be instantiated and called as such:
 * var obj = Object.create(MyHashSet).createNew()
 * obj.add(key)
 * obj.remove(key)
 * var param_3 = obj.contains(key)
 */
let obj = new MyHashSet();
obj.add(1);
obj.add(1);
obj.add(2);
console.log(obj.contains(3));
obj.add(4);
console.log(obj.contains(4));

File: 705/design-hashset.py
class MyHashSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.buckets = 1000
        self.table = [[]] * self.buckets

    def hash(self, key):
        return key % self.buckets

    def pos(self, key):
        return key // self.buckets

    def add(self, key):
        """
        :type key: int
        :rtype: void
        """
        hash_key = self.hash(key)
        if not self.table[hash_key]:
            self.table[hash_key] = [0] * (self.buckets + 1)
        self.table[hash_key][self.pos(key)] = 1

    def remove(self, key):
        """
        :type key: int
        :rtype: void
        """
        hash_key = self.hash(key)
        if self.table[hash_key]:
            self.table[hash_key][self.pos(key)] = 0

    def contains(self, key):
        """
        Returns true if this set did not already contain the specified element
        :type key: int
        :rtype: bool
        """
        hash_key = self.hash(key)
        return (self.table[hash_key] != []) and (self.table[hash_key][self.pos(key)] == 1)


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)
